File: src/Plainion.IronDoc.Tests/XmlDocToMarkdownTests.cs
using System;
using System.IO;
using NUnit.Framework;
using Plainion.IronDoc.FSharp;
using Plainion.IronDoc.Tests.Fakes;

namespace Plainion.IronDoc.Tests
{
    [TestFixture]
    class XmlDocToMarkdownTests
    {
        private XmlDocTransformer myTransformer;
        private XmlDocDocument.Contents myXmlDocumentation;

        [TestFixtureSetUp]
        public void FixtureSetUp()
        {
            var loader = new AssemblyLoader();

            myTransformer = new XmlDocTransformer( loader );

            var assembly = GetType().Assembly;
            var docFile = Path.Combine( Path.GetDirectoryName( assembly.Location ), Path.GetFileNameWithoutExtension( assembly.Location ) + ".xml" );

            myXmlDocumentation = XmlDocDocument.LoadFile( docFile );
        }

        [Test]
        public void SimpleSummary()
        {
            var markdownDocument = Transform( typeof( SimplePublicClass ) );

            Assert.That( markdownDocument, Does.Contain( @"
## Plainion.IronDoc.Tests.Fakes.SimplePublicClass
This is a summary
" ) );
        }

        [Test]
        public void OverwrittenMethods()
        {
            var markdownDocument = Transform( typeof( OverwrittenMethods ) );

            Assert.That( markdownDocument, Does.Contain( @"Returns nicely formatted message about the state of this object" ) );
        }

        [Test]
        public void NestedTypes()
        {
            var markdownDocument = Transform( typeof( NestedType.Nested ) );

            Assert.That( markdownDocument, Does.Contain( @"I am nested" ) );
        }

        private string Transform( Type type )
        {
            using( var writer = new StringWriter() )
            {
                myTransformer.Transform( type, myXmlDocumentation, writer );
                return writer.ToString();
            }
        }
    }
}

File: src/Plainion.IronDoc.Tests/XmlDocToMarkdownTests.fs
﻿namespace Plainion.IronDoc.Tests

open System
open System.IO
open System.Collections.Generic
open NUnit.Framework
open Plainion.IronDoc.FSharp
open Plainion.IronDoc.Tests.Fakes
open Plainion.IronDoc.Tests.Fakes

[<TestFixture>]
module XmlDocToMarkdownTests =
    
    let getApiDoc assemblyLocation =
        let docFile = Path.Combine( Path.GetDirectoryName( assemblyLocation ), Path.GetFileNameWithoutExtension( assemblyLocation ) + ".xml" )

        XmlDocDocument.LoadFile docFile 

    let getApiDocCached = 
        memoize ( fun t -> getApiDoc t )

    let getTransformer () =
        let loader = new AssemblyLoader()
        new XmlDocTransformer( loader )
       
    let getTransformerCached = 
        let transformer = getTransformer
        fun () -> transformer

    let transform ( t : Type ) =
        use writer = new StringWriter()
        
        let apiDoc = getApiDocCached t.Assembly.Location
        let transformer = getTransformer ()
        transformer.TransformType t apiDoc writer
        
        writer.ToString();

    [<Test>]
    let SimpleSummary () = 
        let markdownDocument = transform typedefof<SimplePublicClass>

        Assert.That( markdownDocument, Does.Contain "This is a summary" )

    [<Test>]
    let OverwrittenMethods() =
        let markdownDocument = transform typedefof<OverwrittenMethods>

        Assert.That( markdownDocument, Does.Contain @"Returns nicely formatted message about the state of this object" )

    [<Test>]
    let NestedTypes() =
        let markdownDocument = transform typedefof<NestedType.Nested>

        Assert.That( markdownDocument, Does.Contain @"I am nested" )
